NotificationHubs/src/com/windowsazure/messaging/BaseInstallation.java

```diff
@@ -0,0 +1,310 @@
+package com.windowsazure.messaging;
+
+import com.google.gson.GsonBuilder;
+
+import java.text.SimpleDateFormat;
+import java.util.*;
+
+public abstract class BaseInstallation {
+    private String installationId;
+    private String userId;
+    private NotificationPlatform platform;
+    private boolean pushChannelExpired;
+    private String expirationTime;
+    private List<String> tags;
+    private Map<String, InstallationTemplate> templates;
+    private Map<String, WnsSecondaryTile> secondaryTiles;
+
+    /**
+    * Creates a new installation.
+    */
+    public BaseInstallation() {}
+
+    /**
+     * Creates a new installation with the given installation ID.
+     *
+     * @param installationId The ID for the installation.
+     */
+    public BaseInstallation(String installationId) {
+        this(installationId, (String[]) null);
+    }
+
+    /**
+     * Creates an installation from the Installation ID and tags.
+     *
+     * @param installationId The ID for the installation.
+     * @param tags           The tags for the installation.
+     */
+    public BaseInstallation(String installationId, String... tags) {
+        this(installationId, null, tags);
+    }
+
+    /**
+     * Creates an installation with an installation ID, platform and push channel.
+     *
+     * @param installationId The ID for the installation.
+     * @param platform       The platform for the installation.
+     */
+    public BaseInstallation(String installationId, NotificationPlatform platform) {
+        this(installationId, platform, (String[]) null);
+    }
+
+    public BaseInstallation(String installationId, NotificationPlatform platform, String... tags) {
+        // Validate that this is not FCM
+        validateNotificationPlatform(platform);
+
+        this.installationId = installationId;
+        this.platform = platform;
+        if (tags != null) {
+            for (String tag : tags) {
+                this.addTag(tag);
+            }
+        }
+    }
+
+    /**
+     * Validates that the platform is not FCM. Currently, Notification Hubs supports
+     * FCM Legacy as NotificationPlatform.Gcm Full support for FCM is not currently
+     * supported. See https://aka.ms/AA9dpaz
+     *
+     * @param notificationPlatform The notification platform to verify that it is
+     *                             not FCM.
+     */
+    private static void validateNotificationPlatform(NotificationPlatform notificationPlatform) {
```

Comment:
I'm not a fan of client-side validation, especially when it would get in the way if we add new functionality we want people to be able to take advantage of. :)

Reply:
Also, the Android SDK is built to use tokens generated by FCM, and received on client devices using FCM - so the distinction is pretty narrow: on send operations, which payload format should be submitted to Google? IMHO, that's something that shouldn't be solved or even really thought about at the Installation level. I'd rather the distinction be limited to a header passed during send operations, with different defaults based on the API Version to wean people off of the legacy format.

Reply:
This is necessary because we do not want people to use the FCM platform type as we explicitly say in our README for installations.  We have had several issues filed in this repo because the customer would get a 400 bad request back because we didn't know what "fcm" was on the backend.  This way, we surface it to them at runtime before they make that call.